pallets/proofs-dealer/src/utils.rs

```diff
@@ -257,63 +258,51 @@ where
-                    Some(mutation) if forest_keys_proven.contains(key) => Some((*key, mutation)),
+                    Some(mutation) if forest_keys_proven.contains(key) => {
+                        Some((*key, mutation.clone()))
+                    }
                     Some(_) | None => None,
                 })
                 .collect();
 
             if !mutations.is_empty() {
-                let mut mutations_applied = Vec::new();
-                let new_root = mutations.iter().try_fold(root, |acc_root, mutation| {
-                    // Remove the key from the list of `forest_keys_proven` to avoid having to verify the key proof.
-                    forest_keys_proven.remove(&mutation.0);
-
-                    // Add mutation to list of mutations applied.
-                    mutations_applied.push((mutation.0, mutation.1.clone()));
-
-                    // Apply the mutation to the Forest.
-                    let apply_delta_result = <T::ForestVerifier as TrieProofDeltaApplier<
-                        T::MerkleTrieHashing,
-                    >>::apply_delta(
-                        &acc_root,
-                        &[(mutation.0, mutation.1.clone().into())],
-                        forest_proof,
-                    )
-                    .map_err(|_| Error::<T>::FailedToApplyDelta);
-
-                    // If the mutation was correctly applied, update the Provider's info and return the new root.
-                    match apply_delta_result {
-                        Ok((_, new_root, mutated_keys_and_values)) => {
-                            // Check that the mutated key is the same as the mutation (and is the only one).
-                            ensure!(
-                                mutated_keys_and_values.len() == 1,
-                                Error::<T>::FailedToApplyDelta
-                            );
-                            ensure!(
-                                mutated_keys_and_values[0].0 == mutation.0,
-                                Error::<T>::FailedToApplyDelta
-                            );
-
-                            // Use the interface exposed by the Providers pallet to update the submitting Provider
-                            // after the key removal if the key had a value.
-                            let removed_trie_value = &mutated_keys_and_values[0].1;
-                            if let Some(trie_value) = removed_trie_value {
-                                ProvidersPalletFor::<T>::update_provider_after_key_removal(
-                                    submitter, trie_value,
-                                )
-                                .map_err(|_| Error::<T>::FailedToApplyDelta)?;
-                            }
+                let converted_mutations = mutations
+                    .iter()
+                    .map(|(key, mutation)| (*key, mutation.clone().into()))
+                    .collect::<Vec<(KeyFor<T>, TrieMutation)>>();
```

Comment:
... and then we use that vector (which has a map already) to iterate through it and create another vector. I think this could be done in one single op so not have to iterate twice over the mutations to apply. Just a minor optimization

Reply:
Yeah good catch!